File: frontend/de/unika/ipd/grgen/ast/stmt/graph/ForGraphQueryNode.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 7.1
 * Copyright (C) 2003-2025 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

/**
 * @author Edgar Jakumeit
 */

package de.unika.ipd.grgen.ast.stmt.graph;

import de.unika.ipd.grgen.ast.*;
import de.unika.ipd.grgen.ast.decl.pattern.VarDeclNode;
import de.unika.ipd.grgen.ast.model.type.EdgeTypeNode;
import de.unika.ipd.grgen.ast.model.type.NodeTypeNode;
import de.unika.ipd.grgen.ast.stmt.EvalStatementNode;
import de.unika.ipd.grgen.ast.stmt.NestingStatementNode;
import de.unika.ipd.grgen.ast.type.TypeNode;
import de.unika.ipd.grgen.parser.Coords;

/**
 * AST node representing a for lookup of a neighborhood function.
 */
public abstract class ForGraphQueryNode extends NestingStatementNode
{
	static {
		setName(ForGraphQueryNode.class, "ForGraphQuery");
	}

	BaseNode iterationVariableUnresolved;
	VarDeclNode iterationVariable;

	protected ForGraphQueryNode(Coords coords, BaseNode iterationVariable, CollectNode<EvalStatementNode> loopedStatements)
	{
		super(coords, loopedStatements);
		this.iterationVariableUnresolved = iterationVariable;
		becomeParent(this.iterationVariableUnresolved);
	}

	protected boolean resolveIterationVariable(String forType)
	{
		boolean successfullyResolved = true;

		if(iterationVariableUnresolved instanceof VarDeclNode) {
			iterationVariable = (VarDeclNode)iterationVariableUnresolved;
		} else {
			reportError("Error in resolving iteration variable of for " + forType + " loop.");
			successfullyResolved = false;
		}

		if(!iterationVariable.resolve())
			successfullyResolved = false;

		return successfullyResolved;
	}

	protected boolean checkIterationVariable(String forType)
	{
		TypeNode iterationVariableType = iterationVariable.getDeclType();
		if(!(iterationVariableType instanceof NodeTypeNode)
				&& !(iterationVariableType instanceof EdgeTypeNode)) {
			reportError("Iteration variable of for " + forType + " loop must be of type Node or Edge"
					+ " (but is of type " + iterationVariableType.toStringWithDeclarationCoords() + ").");
			return false;
		}
		
		return true;
	}
}

File: frontend/de/unika/ipd/grgen/ast/stmt/graph/ForGraphQueryNode.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

/// <summary>
/// @author Edgar Jakumeit
/// </summary>

namespace de.unika.ipd.grgen.ast.stmt.graph
{
	using de.unika.ipd.grgen.ast;
	using VarDeclNode = de.unika.ipd.grgen.ast.decl.pattern.VarDeclNode;
	using EdgeTypeNode = de.unika.ipd.grgen.ast.model.type.EdgeTypeNode;
	using NodeTypeNode = de.unika.ipd.grgen.ast.model.type.NodeTypeNode;
	using EvalStatementNode = de.unika.ipd.grgen.ast.stmt.EvalStatementNode;
	using NestingStatementNode = de.unika.ipd.grgen.ast.stmt.NestingStatementNode;
	using TypeNode = de.unika.ipd.grgen.ast.type.TypeNode;
	using Coords = de.unika.ipd.grgen.parser.Coords;

	/// <summary>
	/// AST node representing a for lookup of a neighborhood function.
	/// </summary>
	public abstract class ForGraphQueryNode : NestingStatementNode
	{
		static ForGraphQueryNode()
		{
			SetClassName(typeof(ForGraphQueryNode), "ForGraphQuery");
		}

		internal BaseNode iterationVariableUnresolved;
		internal VarDeclNode iterationVariable;

		protected internal ForGraphQueryNode(Coords coords, BaseNode iterationVariable, CollectNode<EvalStatementNode> loopedStatements)
			: base(coords, loopedStatements)
		{
			this.iterationVariableUnresolved = iterationVariable;
			BecomeParent(this.iterationVariableUnresolved);
		}

		protected internal virtual bool ResolveIterationVariable(string forType)
		{
			bool successfullyResolved = true;

			if(iterationVariableUnresolved is VarDeclNode)
				iterationVariable = (VarDeclNode)iterationVariableUnresolved;
			else
			{
				ReportError("Error in resolving iteration variable of for " + forType + " loop.");
				successfullyResolved = false;
			}

			if(!iterationVariable.Resolve())
				successfullyResolved = false;

			return successfullyResolved;
		}

		protected internal virtual bool CheckIterationVariable(string forType)
		{
			TypeNode iterationVariableType = iterationVariable.DeclType;
			if(!(iterationVariableType is NodeTypeNode)
					&& !(iterationVariableType is EdgeTypeNode))
			{
				ReportError("Iteration variable of for " + forType + " loop must be of type Node or Edge"
						+ " (but is of type " + iterationVariableType.ToStringWithDeclarationCoords() + ").");
				return false;
			}

			return true;
		}
	}

}
